sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/schemas/AwsTypes.java

```diff
@@ -0,0 +1,295 @@
+    private static class ToAws extends ConverterFactory {
+      private final Factory<SerializableFunction<Row, ?>> fromRowFactory;
+
+      ToAws(Factory<SerializableFunction<Row, ?>> fromRowFactory) {
+        super(AwsTypes::toJavaInstant, AwsTypes::toSdkBytes, true);
+        this.fromRowFactory = fromRowFactory;
+      }
+
+      @Override
+      @SuppressWarnings("nullness") // schema nullable for this factory
+      protected SerializableFunction pojoTypeConverter(SdkField<?> field) {
+        return fromRowFactory.create(targetClassOf(field.constructor().get()), null);
+      }
+    }
+
+    /**
+     * Converter factory from AWS types to Beam raw unmodified row types. This is applicable for
+     * getters and also removes default values for lists & maps to avoid serializing those.
+     */
+    private static class FromAws extends ConverterFactory {
+      private static final ConverterFactory INSTANCE = new FromAws();
+
+      FromAws() {
+        super(AwsTypes::toJodaInstant, AwsTypes::toBytes, false);
+      }
+
+      @Override
+      SerializableFunction create(SerializableFunction fn, SdkField<?> field) {
+        MarshallingType<?> type = field.marshallingType();
+        if (type.equals(MAP)) {
+          fn = skipDefaultMap(fn);
+        } else if (type.equals(LIST)) {
+          fn = skipDefaultList(fn);
+        }
+        return needsConversion(field) ? super.create(fn, field) : fn;
+      }
+
+      @SuppressWarnings("nullness")
+      private static SerializableFunction skipDefaultList(SerializableFunction fn) {
+        return in -> {
+          Object list = fn.apply(in);
+          return list != DefaultSdkAutoConstructList.getInstance() ? list : null;
+        };
+      }
+
+      @SuppressWarnings("nullness")
+      private static SerializableFunction skipDefaultMap(SerializableFunction fn) {
+        return in -> {
+          Object map = fn.apply(in);
+          return map != DefaultSdkAutoConstructMap.getInstance() ? map : null;
+        };
+      }
+    }
+  }
+
+  // Convert upper camel SDK field names to lower camel
+  static String normalizedNameOf(SdkField<?> field) {
+    String name = field.memberName();
+    return name.length() > 1 && Ascii.isLowerCase(name.charAt(1))
+        ? Ascii.toLowerCase(name.charAt(0)) + name.substring(1)
+        : name.toLowerCase(Locale.ROOT);
+  }
+
+  static java.time.Instant toJavaInstant(Object instant) {
+    return java.time.Instant.ofEpochMilli(((Instant) instant).getMillis());
+  }
+
+  private static Instant toJodaInstant(Object instant) {
+    return Instant.ofEpochMilli(((java.time.Instant) instant).toEpochMilli());
+  }
+
+  private static SdkBytes toSdkBytes(Object sdkBytes) {
+    return SdkBytes.fromByteArrayUnsafe((byte[]) sdkBytes); // TODO copy or use unsafe?
+  }
+
+  private static byte[] toBytes(Object sdkBytes) {
+    return ((SdkBytes) sdkBytes).asByteArrayUnsafe(); // TODO copy or use unsafe?
```

Comment:
Are these TODOs something we can address now, or should there be an issue tracking it?

Reply:
I had a look at what GetterBasedSchemaProvider does for `byte[]` / `ByteBuffer`. It's exposing the mutable bytes as well, so I'll keep the unsafe operations here. 👍